src/libraries/System.Net.Primitives/src/System/Net/IPAddressParser.cs

```diff
@@ -65,30 +69,54 @@ private static unsafe bool TryParseIpv4(ReadOnlySpan<char> ipSpan, out long addr
             return false;
         }
 
-        private static unsafe bool TryParseIPv6(ReadOnlySpan<char> ipSpan, Span<ushort> numbers, int numbersLength, out uint scope)
+        private static unsafe bool TryParseIPv6<TChar>(ReadOnlySpan<TChar> ipSpan, Span<ushort> numbers, int numbersLength, out uint scope)
+            where TChar : unmanaged, IBinaryInteger<TChar>
         {
+            Debug.Assert(typeof(TChar) == typeof(char) || typeof(TChar) == typeof(byte));
             Debug.Assert(numbersLength >= IPAddressParserStatics.IPv6AddressShorts);
 
             int end = ipSpan.Length;
-
             bool isValid = false;
-            fixed (char* ipStringPtr = &MemoryMarshal.GetReference(ipSpan))
+            fixed (TChar* ipStringPtr = &MemoryMarshal.GetReference(ipSpan))
             {
                 isValid = IPv6AddressHelper.IsValidStrict(ipStringPtr, 0, ref end);
             }
+            scope = 0;
             if (isValid || (end != ipSpan.Length))
             {
-                string? scopeId = null;
-                IPv6AddressHelper.Parse(ipSpan, numbers, 0, ref scopeId);
+                IPv6AddressHelper.Parse(ipSpan, numbers, out ReadOnlySpan<TChar> scopeIdSpan);
 
-                if (scopeId?.Length > 1)
+                if (scopeIdSpan.Length > 1)
                 {
-                    if (uint.TryParse(scopeId.AsSpan(1), NumberStyles.None, CultureInfo.InvariantCulture, out scope))
+                    string interfaceName = string.Empty;
+                    scopeIdSpan = scopeIdSpan.Slice(1);
+
+                    // scopeId is a numeric value
+                    if (typeof(TChar) == typeof(byte))
+                    {
+                        ReadOnlySpan<byte> castScopeIdSpan = MemoryMarshal.Cast<TChar, byte>(scopeIdSpan);
+
+                        if (uint.TryParse(castScopeIdSpan, NumberStyles.None, CultureInfo.InvariantCulture, out scope))
+                        {
+                            return true;
+                        }
+
+                        interfaceName = System.Text.Encoding.UTF8.GetString(castScopeIdSpan);
```

Comment:
We shouldn't be allocating here or below (`interfaceName = new string(castScopeIdSpan);`). We should ensure the path through InterfaceNameToIndex can take spans.

Reply:
InterfaceNameToIndex is the entry point to the OS-specific PInvoke paths; there's no managed code (other than the library call) within that method. I have a commit which can feed the spans through PInvoke into the underlying native libraries without constructing a string, but it's more involved than the rest of the PR and I want to make sure I'm not about to re-inflate the diff or widen the PR scope over a misunderstanding.

The underlying if_nametoindex API works with a null-terminated ANSI byte array on Windows and (indirectly) a null-terminated UTF8 array on Linux. There are marshallers (AnsiStringMarshaller, Utf8StringMarshaller) which convert from strings on Windows, but they don't handle the two kinds of Span we care about. I've thus got two new internal marshallers (AnsiReadOnlySpanMarshaller, Utf8ReadOnlySpanMarshaller) which replicate these.

To prove these marshallers, I've added a test which round-trips each interface's index to and from an interface name (via a string, Span<char> and Span<byte>). This is fine on Windows, but means adding a method to System.Native on Linux. By the time the marshallers, test and change to System.Native are done, I've nearly doubled the size of the PR.

Is this the type of change you want in order to avoid the string allocation, or am I misunderstanding?

I'd personally prefer to expand the existing string marshallers (AnsiStringMarshaller, BStrStringMarshaller, Utf16StringMarshaller, Utf8StringMarshaller) so they can marshal the Span/ReadOnlySpan variations we care about. This would eliminate ~90% of the changes I'm describing, and seamless marshalling of spans into native libraries would be helpful more broadly. I'm pretty sure it'd need an API review though.

Reply:
It shouldn't need full-fledged reusable marshalers as part of this change. It can just have the helpers necessary to do the one-off marshaling, e.g. instead of:
```C#
[LibraryImport(Libraries.SystemNative, EntryPoint = "SystemNative_InterfaceNameToIndex", StringMarshalling = StringMarshalling.Utf8, SetLastError = true)]
public static partial uint InterfaceNameToIndex(string name);
```
it can be something relatively simple like:
```C#
public static partial uint InterfaceNameToIndex(ReadOnlySpan<byte> utf8Name)
{
    byte* ptr = (byte*)NativeMemory.Alloc((nuint)utf8Name.Length + 1);
    try
    {
        utf8Name.CopyTo(new Span<byte>(ptr, utf8Name.Length));
        ptr[utf8Name.Length] = 0;
        return InterfaceNameToIndex(ptr);
    }
    finally
    {
        NativeMemory.Free(ptr);
    }
}

public static partial uint InterfaceNameToIndex(ReadOnlySpan<char> name)
{
    int byteCount = Encoding.UTF8.GetByteCount(name);
    byte* ptr = (byte*)NativeMemory.Alloc((nuint)byteCount + 1);
    try
    {
        Encoding.UTF8.GetBytes(name, new Span<byte>(ptr, byteCount));
        ptr[byteCount] = 0;
        return InterfaceNameToIndex(ptr);
    }
    finally
    {
        NativeMemory.Free(ptr);
    }
}

[LibraryImport(Libraries.SystemNative, EntryPoint = "SystemNative_InterfaceNameToIndex", SetLastError = true)]
private static partial uint InterfaceNameToIndex(byte* utf8NullTerminatedName);
```

Reply:
I understand, thanks. I was trying to encapsulate the marshalling code by using a dedicated marshaller, but that might be overkill. I'll use a simpler one-off marshalling method here, think more about those expansions to the string marshallers, then come back and refactor if all goes well.